Using Pinned Hashes for JEDI Repositories

asewnath opened the pull request:
## Description
Adds two options for using pinned versions in swell experiments:

1. Adds `pinned_create` jedi build method. Calling this method will clone jedi repositories with hashes in `pinned_versions.yaml`. Then it will build jedi and proceed as normal.
2. Adds `use_pinned_existing` jedi build method. This will point to `build` and `source` directories in `/discover/nobackup/projects/gmao/advda/jedi_bundles` or `/discover/nobackup/projects/gmao/advda/jedi_bundles_sles15` for milan nodes. This build is currently pinned to August 31st. The `check_hashes` script is called to check if commit hashes in this build lines up with `pinned_versions.yaml`. If it does not, it will abort. If it does, it will proceed as normal.
3. Quick change for edit documents link

Note that none of these options are default and they must be indicated through the CLI.

### Notes
- Automatically building jedi in local directory if `use_pinned_existing` fails will be addressed in another PR
- Github action will be addressed in different PR

## Issues
Resolves #423
Addresses some concerns in #345

asewnath committed:
suppress abort message and add check hash utility

asewnath committed:
code changes

asewnath committed:
Merge branch 'develop' into feature/pinned_versions_utils

asewnath committed:
adding ability to use pinned hashes

asewnath committed:
merge fix

asewnath commented on src/swell/tasks/build_jedi.py:
```diff
@@ -33,7 +33,8 @@ def execute(self) -> None:
         # ----------------------------------------------------------------
-        if self.config.jedi_build_method() == 'create':
+        if self.config.jedi_build_method() == 'create' or \
+           self.config.jedi_build_method() == 'pinned_create':
```
```suggestion
           self.config.jedi_build_method() in ('create', 'pinned_create')
```

asewnath committed:
code change

asewnath committed:
update edit document link on docs

asewnath committed:
updating docs

asewnath committed:
swell test fix

asewnath committed:
Merge branch 'feature/pinned_versions_support' of https://github.com/GEOS-ESM/swell into feature/pinned_versions_support

asewnath committed:
Merge branch 'develop' into feature/pinned_versions_support

asewnath committed:
Merge branch 'develop' into feature/pinned_versions_support

asewnath committed:
fixing swell test

asewnath committed:
check hash platform test

asewnath committed:
Update test_ci_application_discover.yml

asewnath committed:
Update test_ci_application_discover.yml

asewnath committed:
Update test_ci_application_discover.yml

asewnath committed:
Update tier1_application_discover.yml

asewnath committed:
Update tier1_application_discover.yml

asewnath committed:
fix code test

ashiklom commented on src/swell/tasks/build_jedi.py:
```diff
@@ -65,7 +70,7 @@ def execute(self) -> None:
             self.logger.abort(f'Found \'{self.config.jedi_build_method()}\' for ' +
                               f'jedi_build_method in the  experiment dictionary. Must be ' +
-                              f'\'create\'.')
+                              f'\'create\' or \'pinned_create\'.')
```
Align the happy path to the left edge! A.k.a., Return early!

In general, rather than really large if-else blocks like this...

```python
if good_condition:
  # do...
  # a...
  # bunch...
  # of...
  # stuff...
else: 
  raise SomeError(...)
```

a much better approach is something like:

```python
if not good_condition:
  raise SomeError()

# Do...
# a...
# bunch...
# of...
# stuff.
# ^^ Note: This block does *not* have to be indented anymore!
```

Here, I would flip the conditional to `if jedi_build_method not in ('create', 'pinned_create')`, put the error there, and then unindent the rest of the code (the "happy path").

(I know this isn't your code, but can't hurt to make cheap style improvements when we can 😄 ).

asewnath replied:
TIL what happy path is 🥲

ashiklom commented on src/swell/test/platform_tests/__init__.py:
```diff
@@ -0,0 +1,9 @@
+
+import os
+
+repo_directory = os.path.dirname(__file__)
```
An `__init__.py` file is required for Python to pick this up as a module. But you don't actually need anything in the `__init__.py` file.

I did a quick scan and we never actually use the `repo_directory` variable anywhere (and, even if we did, the more correct thing to do would be to use `importlib.resources` anyway).

At some point, we should go through and remove all this unnecessary code. But, in the meantime, we can at least stop adding new instances of it.

```suggestion
```

ashiklom commented on src/swell/utilities/build.py:
```diff
@@ -57,6 +60,7 @@ def set_jedi_bundle_config(
     path_to_source: str,
     path_to_build: str,
     platform: str,
+    use_pinned: bool,
```
I don't necessarily disagree that we should include this as a required argument --- it has a major impact on what Swell does, so it makes sense to force users to specify this.

However, just noting that introducing new _required_ arguments to things is a breaking change, and, in general, should be considered carefully and communicated accordingly.

We can keep this as is, but just flagging it for future consideration.

asewnath replied:
Sure, I can make changes like this more obvious in the PR description. The flag communicates to `set_jedi_bundle_config` that it needs to create it's jedi bundle build config with the pinned versions that swell tracks. If there's a better way of doing this please let me know

ashiklom replied:
TLDR: No changes requested here; this can be resolved 😄 

I think this is fine. It's also internal code, so it's somewhat less important than a change to the exposed user interface (though, note that unlike some other languages like Java, Python has no formal concept of public vs. private methods/attributes and manages these things purely by convention and best practices. There's no such thing as a truly private Python method in the same way as there is in Java, so you never know who is relying on your internal implementation details!).

I wouldn't handle this any differently than you did. I was just making a general comment that changing the _required_ part of a function signature is a common kind of breaking change. Breaking changes are fine, especially in relatively early phases of software development. Enforcing strict backwards compatibility too early can lock us into suboptimal design choices.

That said, what we need to do at _some_ point with Swell is adopt proper [Semantic Versioning](https://semver.org/), possibly paired with something like [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/). Since we squash-merge everything anyway, we would only have to enforce something like Conventional Commits at the PR level rather than at the individual commit level.

asewnath replied:
Thanks for the information!

ashiklom commented on src/swell/tasks/clone_jedi.py:
```diff
@@ -40,7 +41,15 @@ def execute(self) -> None:
             link_path(self.config.existing_jedi_source_directory(), jedi_bundle_source_path)
 
-        elif self.config.jedi_build_method() == 'create':
+        elif self.config.jedi_build_method() == 'use_pinned_existing':
```
Somewhat related to my earlier comment about the "happy path", it may be a good idea to put the check for valid values of the `jedi_build_method` early and then, separately, to go through the conditional execution. I.e., Something like:

```python
build_method = self.config.jedi_build_method()

if not build_method in ("use_existing", "use_pinned_existing", "create", "pinned_create"):
  logger.abort("...")

# Now, proceed with the code:
if build_method == "use_existing":
  ...
elif ... # and so on
```

This makes it very easy to quickly for other developers to see what all the valid values of `build_method` are.

Since this pattern occurs in many places, an even better thing to do would be to define a `JEDI_VALID_BUILD_METHODS` constant in _one_ place somewhere and then check `jedi_build_method` against that.

Even better is to do this check as part of the `self.config` initialization. Define which values of `jedi_build_method` are valid, and then throw an error immediately if a user tries to set that to anything else. Then, we don't have to keep redundantly checking that this is valid everywhere else in the Swell code.

For inspiration, I suggest reading this: [Parse, don't validate](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/). The language there is Haskell, but the underlying idea --- use the most precise data type you can --- is universal.

ashiklom replied:
This idea is also why I think Python's dataclasses (and Pydantic data models) are such a good idea. With dataclasses, we can specify valid values of `self.config.jedi_build_method()` through an `Enum` type annotation (though that won't do any runtime checking). With Pydantic, we do that too, but we _also_ get runtime checking and clear errors built in (though we have to eat the cost of having another external dependency).

ashiklom reviewed (changes requested):
Nice work!

In addition to the stuff in my review, one more requested change (I couldn't add a suggestion because it's too far from any code you touched): In `src/swell/tasks/task_questions.yaml` line 466 (definition of `jedi_build_method`), we need to add the new valid options.

Also, if we do that, is there a way to _enforce_ a particular set of options at runtime? I.e., When you go to set `self.config.jedi_build_method`, if it's set to something invalid, it'll fail immediately? Not sure if the `string-drop-list` or some other `type` will do that for us. In any case, if there's a way to easily do that, we should do it, because that will allow us to remove lots of `if self.config.jedi_build_method() in (...)` code blocks.

Dooruk reviewed (dismissed):
Thanks, looks good!

We just need to ensure the consistency of `pinned_versions.yaml` and the following locations:

https://github.com/GEOS-ESM/swell/blob/597bbbe9c87867178a130a10c6d07418d1a212d8/src/swell/deployment/platforms/nccs_discover_sles15/task_questions.yaml#L10-L14

asewnath committed:
adding task question changes

asewnath committed:
Merge branch 'develop' into feature/pinned_versions_support

asewnath committed:
Merge branch 'feature/pinned_versions_support' of https://github.com/GEOS-ESM/swell into feature/pinned_versions_support

asewnath committed:
swell test

asewnath commented:
> Thanks, looks good!
> 
> We just need to ensure the consistency of `pinned_versions.yaml` and the following locations:
> 
> https://github.com/GEOS-ESM/swell/blob/597bbbe9c87867178a130a10c6d07418d1a212d8/src/swell/deployment/platforms/nccs_discover_sles15/task_questions.yaml#L10-L14

I modified the platform task_question.yaml files to point to pinned version jedi bundles for SLES12 and SLES15.

asewnath committed:
refactor: addressing PR comments

asewnath committed:
refactor: removing contents of init files in test directory

asewnath committed:
refactor(tasks): modifying build_jedi_by_linking for clarity

asewnath commented:
> Also, if we do that, is there a way to _enforce_ a particular set of options at runtime? I.e., When you go to set `self.config.jedi_build_method`, if it's set to something invalid, it'll fail immediately? Not sure if the `string-drop-list` or some other `type` will do that for us. In any case, if there's a way to easily do that, we should do it, because that will allow us to remove lots of `if self.config.jedi_build_method() in (...)` code blocks.

In `task_questions.yaml`, I modified `jedi_bundle_method` to include the new valid options,  and then I changed it to `string-drop-list` which works well for constraining the user to only valid build methods. If there are any remaining `if self.config.jedi_build_method() in (...)` code blocks, it's because the task itself is checking for specific build methods and aborts if other valid methods were provided. For example, `BuildJediByLinking` aborts if the user has selected either the `create` or `pinned_create` build method.

ashiklom commented on src/swell/tasks/clone_jedi.py:
```diff
@@ -37,20 +37,16 @@ def execute(self) -> None:
 
-        elif self.config.jedi_build_method() == 'use_pinned_existing':
-
+        if self.config.jedi_build_method() == 'use_pinned_existing':
```
This one should probably be `elif`, right? Since we're _either_ linking `jedi_bundle_source_path` to the pinned source directory _or_ the existing source directory.

asewnath replied:
`CloneJedi` also handles the other jedi build methods: `if self.config.jedi_build_method() in ('create', 'pinned_create')`. I'm not sure what the `else` statement would be after all the `elif` statements since we are assuming the user passes a valid jedi build method

ashiklom replied:
So I think this is a matter of style. `if + elif` explicitly communicates that the statements are related and the underlying clauses are mutually exclusive --- "either X or Y or Z".

https://stackoverflow.com/questions/22782785/when-to-use-if-vs-elif-in-python

My recommendation would be to use `elif` here to clearly communicate that `jedi_build_method` can only be one of those things and these are mutually exclusive clauses.

You don't need a final `else` statement --- it's perfectly valid to have a bunch of `elif` statements without a concluding `else`.

asewnath committed:
refactor(tasks): updating CloneJedi to use elif
